core/base/arrayPreconditioning/ArrayPreconditioning.h

```diff
@@ -0,0 +1,593 @@
+///
+
+#pragma once
+
+// ttk common includes
+#include <Debug.h>
+
+#ifdef TTK_ENABLE_MPI
+#include <mpi.h>
+#endif
+
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+namespace ttk {
+
+  /**
+   * The ArrayPreconditioning class provides methods to generate order arrays
+   * from a selection of scalar field arrays.
+   */
+  class ArrayPreconditioning : virtual public Debug {
+
+  public:
+    ArrayPreconditioning();
+#ifdef TTK_ENABLE_MPI
+    MPI_Datatype getMPIType(const float ttkNotUsed(val)) const {
+      return MPI_FLOAT;
+    };
+    MPI_Datatype getMPIType(const int ttkNotUsed(val)) const {
+      return MPI_INT;
+    };
+    MPI_Datatype getMPIType(const unsigned int ttkNotUsed(val)) const {
+      return MPI_UNSIGNED;
+    };
+    MPI_Datatype getMPIType(const double ttkNotUsed(val)) const {
+      return MPI_DOUBLE;
+    };
+    MPI_Datatype getMPIType(const long double ttkNotUsed(val)) const {
+      return MPI_LONG_DOUBLE;
+    };
+    MPI_Datatype getMPIType(const long ttkNotUsed(val)) const {
+      return MPI_LONG;
+    };
+    MPI_Datatype getMPIType(const unsigned long ttkNotUsed(val)) const {
+      return MPI_UNSIGNED_LONG;
+    };
+    MPI_Datatype getMPIType(const long long ttkNotUsed(val)) const {
+      return MPI_LONG_LONG;
+    };
+    MPI_Datatype getMPIType(const unsigned long long ttkNotUsed(val)) const {
+      return MPI_UNSIGNED_LONG_LONG;
+    };
+#endif
+    template <typename DT, typename IT>
+    struct value {
+      DT scalar;
+      IT globalId;
+      IT localId;
+      IT ordering = -1;
+
+      value(DT _scalar, IT _globalId, IT _localId)
+        : scalar(_scalar), globalId(_globalId), localId(_localId) {
+      }
+    };
+
+    // creates a vector of value structs based on the given pointers
+    // only takes values of vertices which mainly belong to the current rank
+    // ( so only vertices which are no ghost cells)
+    template <typename DT, typename IT>
+    void populateVector(std::vector<value<DT, IT>> &valuesToSortVector,
+                        std::vector<IT> &gidsToGetVector,
+                        std::unordered_map<IT, IT> &gidToLidMap,
+                        const size_t nVerts,
+                        const DT *const scalars,
+                        const IT *const globalIds,
+                        const char *const ghostCells) const {
+      for(size_t i = 0; i < nVerts; i++) {
+        IT globalId = globalIds[i];
+        IT localId = i;
+        gidToLidMap[globalId] = localId;
+        if((int)ghostCells[i] == 0) {
+          float scalarValue = scalars[i];
+          valuesToSortVector.emplace_back(scalarValue, globalId, localId);
+        } else {
+          gidsToGetVector.push_back(globalId);
+        }
+      }
+    }
+
+    // orders an value vector first by their scalar value and then by global id
+    template <typename DT, typename IT>
+    void sortVerticesDistributed(std::vector<value<DT, IT>> &values) const {
+      std::sort(
```

Comment:
this is a sequential sort. 
please use OpenMp if available (see `core/base/common/OrderDisambiguation.h`:78, function `sortVertices()`)